src/core/qgssqlexpressioncompiler.cpp

```diff
@@ -307,6 +307,13 @@ QgsSqlExpressionCompiler::Result QgsSqlExpressionCompiler::compileNode( const Qg
         return Fail;
     }
 
+    case QgsExpressionNode::ntBetweenOperator:
+    {
+      const QgsExpressionNodeBetweenOperator *n = static_cast<const QgsExpressionNodeBetweenOperator *>( node );
+      result = n->dump();
```

Comment:
Are you confident that all subclasses of QgsSqlExpressionCompiler support between? (i.e. sqlite, hana, sql server, postgres, spatialite, oracle)

Reply:
We'll need a test added to feature source test base to cover this -- somewhere here: https://github.com/qgis/QGIS/blob/master/tests/src/python/featuresourcetestbase.py#L161

Reply:
I'm pretty confident that BETWEEN is widely supported by all SQL dialects, I checked postgres and OGR but I'll add a test as you recommend.

Reply:
@nyalldawson I'm making progresses with the iterator tests but there is a wierd issue with datetime on spatialite provider (works fine when connecting through OGR though, IIRC there was some logic in OGR to handle this Zulu issue):

the expression is compiled as
`select * from somedata where dt BETWEEN '2020-05-03T12:13:14Z' AND '2020-05-04T12:14:14Z'`

but this does not match the test data `2020-05-03T12:13:14` (note: no `Z`ulu), column type for `dt` is `TIMESTAMP`, any idea?

Reply:
Hmm, not sure sorry. I do know that sqlite datetime handling is completely horrible though. If you're struggling with getting this working well I'd suggest just blocking the compilation of this node type for the spatialite provider (better slow and safe vs fast but inaccurate)

Reply:
I could skip the between operator, but this seems to be a more general issue: this works

`select * from somedata where dt = '2020-05-03T12:13:14'`

this doesn't

`select * from somedata where dt = '2020-05-03T12:13:14Z'`

It seems like the `Z` comes from here:
https://github.com/qgis/QGIS/blob/master/src/core/qgssqliteexpressioncompiler.cpp#L130

The docs here say `Z` is a no-op https://www.sqlite.org/lang_datefunc.html but it doesn't seem to be the case.

Try to remove `Z` now ...

Reply:
> Try to remove Z now ...

Oh, please **don't** do that. It's likely to break the most important use case of date time compilation we have -- which is to ensure that datetime filters on gpkg work correctly and take advantage of attribute indices. I'd FAR rather "between" is never compiled for sqlite sources then see that regress. 

Ping @nirvn, who may have further insights on sqlite/gpkg date handling

Reply:
Yeah, it was only a test, but the problem is not limited to BETWEEN as you can see from the `=` sample query above.

Reply:
Here is a full session to demonstrate the issue:

```sql
@blackhole ~/dev/QGIS (between-expression)$ sqlite3 /home/dev/QGIS/tests/testdata/provider/spatialite.db
SQLite version 3.31.1 2020-01-27 19:55:54
Enter ".help" for usage hints.
sqlite> select cnt, dt from somedata; 
100|2020-05-03T12:13:14
200|2020-05-04T12:14:14
300|
400|2021-05-04T13:13:14
-200|2020-05-04T12:13:14
sqlite> select cnt, dt from somedata where dt = '2020-05-03T12:13:14Z';
sqlite> select cnt, dt from somedata where dt = '2020-05-03T12:13:14';
100|2020-05-03T12:13:14
```